library/std/src/net/ip.rs

```diff
@@ -3,12 +3,44 @@
 mod tests;
 
 use crate::cmp::Ordering;
-use crate::fmt::{self, Write as FmtWrite};
-use crate::io::Write as IoWrite;
+use crate::fmt::{self, Write};
 use crate::mem::transmute;
+use crate::str;
 use crate::sys::net::netc as c;
 use crate::sys_common::{FromInner, IntoInner};
 
+/// Used for slow path in `Display` implementations when alignment is required.
+struct IpDisplayBuffer<const SIZE: usize> {
+    buf: [u8; SIZE],
+    len: usize,
+}
+
+impl<const SIZE: usize> IpDisplayBuffer<SIZE> {
+    #[inline(always)]
+    pub const fn new(_ip: &[u8; SIZE]) -> Self {
+        Self { buf: [0; SIZE], len: 0 }
+    }
+
+    #[inline(always)]
+    pub fn as_str(&self) -> &str {
+        // SAFETY: `buf` is only written to by the `fmt::Write::write_str` implementation
+        // which writes a valid UTF-8 string to `buf` and correctly sets `len`.
+        unsafe { str::from_utf8_unchecked(&self.buf[..self.len]) }
+    }
+}
+
+impl<const SIZE: usize> fmt::Write for IpDisplayBuffer<SIZE> {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        if let Some(buf) = self.buf.get_mut(self.len..(self.len + s.len())) {
```

Comment:
Note: There is no overflow bug here because if it overflows the end will be less than beginning and `get_mut` will just return `None` which is intended.